src/nvim/mbyte.h

```diff
@@ -59,3 +62,62 @@ static inline int mb_strcmp_ic(bool ic, const char *s1, const char *s2)
 // multi-byte characters if needed. Only use with "p" > "s" !
 #define MB_PTR_BACK(s, p) \
   (p -= utf_head_off((char *)(s), (char *)(p) - 1) + 1)
+
+static inline CharInfo ptr2CharInfo(char const *p_in)
+  REAL_FATTR_NONNULL_ALL REAL_FATTR_PURE REAL_FATTR_WARN_UNUSED_RESULT REAL_FATTR_ALWAYS_INLINE;
+
+/// Convert a UTF-8 byte sequence to a Unicode code point.
+/// Handles ascii, multibyte sequiences and illegal sequences.
+///
+/// @param[in]  p_in  String to convert.
+///
+/// @return information abouth the character. When the sequence is illegal,
+/// 'value' is negative, 'len' is 1.
+static inline CharInfo ptr2CharInfo(char const *const p_in)
```

Comment:
this publicly visible function should be called `utf_ptr2CharInfo`. The internal helper function should be named something which indicated that (`utf_ptr2CharInfo_impl` is a pattern we have used elsewhere for implementation functions which are not meant to be called directly ).